pkg/controller/deactivation/deactivation_controller.go

```diff
@@ -149,10 +165,47 @@ func (r *ReconcileDeactivation) Reconcile(request reconcile.Request) (reconcile.
 	logger.Info("user account time values", "deactivation timeout duration", deactivationTimeout, "provisionedTimestamp", provisionedTimestamp)
 
 	timeSinceProvisioned := time.Since(provisionedTimestamp.Time)
-	if timeSinceProvisioned < deactivationTimeout {
+
+	deactivatingNotificationTimeout := time.Duration((deactivationTimeoutDays-config.Deactivation.DeactivatingNotificationDays)*24) * time.Hour
+
+	if timeSinceProvisioned < deactivatingNotificationTimeout {
+		// It is not yet time to send the deactivating notification so requeue until it will be time to send it
+		requeueAfterTimeToNotify := deactivatingNotificationTimeout - timeSinceProvisioned
+		logger.Info("requeueing request", "RequeueAfter", requeueAfterTimeToNotify,
+			"Expected deactivating notification date/time", time.Now().Add(requeueAfterTimeToNotify).String())
+		return reconcile.Result{RequeueAfter: requeueAfterTimeToNotify}, nil
+	}
+
+	// If the usersignup state hasn't been set to deactivating, then set it now
+	if !states.Deactivating(usersignup) {
+		states.SetDeactivating(usersignup, true)
+
+		logger.Info("setting usersignup state to deactivating")
+		if err := r.client.Update(context.TODO(), usersignup); err != nil {
+			logger.Error(err, "failed to update usersignup")
+			return reconcile.Result{}, err
+		}
+
+		// Requeue so that the deactivation due time can be calculated after the notification has been sent
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	deactivatingCondition, found := condition.FindConditionByType(usersignup.Status.Conditions,
+		toolchainv1alpha1.UserSignupUserDeactivatingNotificationCreated)
+	if !found || deactivatingCondition.Reason != toolchainv1alpha1.UserSignupDeactivatingNotificationCRCreatedReason {
+		// If the UserSignup has been marked as deactivating, however the deactivating notification hasn't been
+		// created yet, then requeue - the notification should be created shortly
+		return reconcile.Result{Requeue: true}, nil
```

Comment:
First. I think it's worth mentioning that the notification is created by the UserSignup controller when we set the state to "deactivating".
Second. I'm concerned about always requeuing without any delay if the notification condition is not in the expected state. What if there is some global problem with notification creation? Will we "DDoS" API server by requeuing requests in this case?
Should we either add some delay or even return an error so, the controller manager takes care of properly delaying the reconcile requests?
WDYT?

Reply:
No problem, I've updated the comment as such.  As I mentioned above, it might be worth (and certainly wouldn't hurt) to put in a small delay between reconciles.  I suggested 5 seconds, however there's really no reason why this couldn't be much more conservative, even as high as 10 minutes wouldn't hurt here in my opinion.

Reply:
It's deactivation flow which normally happens once in 30 days. So, I think a 10 second delay would be totally fine :)

Reply:
Adjusted delay to 10 seconds.